src/openforms/contrib/haal_centraal/clients/__init__.py

```diff
@@ -1,25 +1,67 @@
-from zgw_consumers_ext.api_client import ServiceClientFactory
+from typing import Any, cast
 
-from ..models import HaalCentraalConfig
+from openforms.authentication.constants import AuthAttribute
+from openforms.config.models import GlobalConfiguration
+from openforms.submissions.models import Submission
+from zgw_consumers_ext.api_client import build_client
+
+from ..constants import DEFAULT_X_GEBRUIKER_HEADER
+from ..models import HaalCentraalConfig, HaalCentraalRequestOptions
 from .brp import CLIENT_CLS_FOR_VERSION as BRP_CLIENT_CLS_FOR_VERSION, BRPClient
 
 
 class NoServiceConfigured(RuntimeError):
     pass
 
 
-def get_brp_client(**kwargs) -> BRPClient:
+def get_brp_client(submission: Submission | None = None, **kwargs: Any) -> BRPClient:
+    """Get an instance of a BRP client.
+
+    :param submission: the current submission in the context of the client usage.
+        This submission will be used to fetch headers overrides.
+    :param kwargs: additional kwargs to be passed to the client class.
+    """
+
     config = HaalCentraalConfig.get_solo()
+    global_config = GlobalConfiguration.get_solo()
     assert isinstance(config, HaalCentraalConfig)
-    if not (service := config.brp_personen_service):
+    assert isinstance(global_config, GlobalConfiguration)
+    if not config.brp_personen_service:
         raise NoServiceConfigured("No BRP service configured!")
 
+    origin_oin = global_config.organization_oin
+    doelbinding = config.default_x_doelbinding_header
+    verwerking = config.default_x_verwerking_header
+    gebruiker = DEFAULT_X_GEBRUIKER_HEADER
+
+    if submission is not None:
+        request_options = cast(
+            HaalCentraalRequestOptions | None,
+            getattr(submission.form, "haal_centraal_request_options", None),
+        )
```

Comment:
`assert isinstance(object, "tuple_or_union_type")` is not a construct understood by type checkers to this day, so using it would defeat the purpose of the current type casting

Reply:
cool - didn't know about the type checker limitations, so if it pops up again, add a comment describing why `cast` is used instead of `assert` :+1:

Reply:
Yes I was actually surprised by it too